### PR TITLE
[IOTDB-4126] Optimize cache implementation using HashMap to find Node instead of traversal with a for loop

### PR DESCRIPTION
**Pull request by ZhanGHanG9991**
## Description
Optimize cache implementation using HashMap to find Node instead of traversal with a for loop

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/Cache.java [hunk 1]
```diff
@@ -64,33 +69,37 @@ protected Cache(int capacity) {
...
     Node last = tail.previous.remove();
     cachedNodes[--cacheSize] = last;
+    valueToNode.remove(last.value);
     return last.value;
   }
 
   protected void addFirst(int value) {
-    cachedNodes[cacheSize++].set(head, value, head.succeeding);
+    cachedNodes[cacheSize].set(head, value, head.succeeding);
+    valueToNode.put(value, cachedNodes[cacheSize]);
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/Cache.java:95]: Does this work well when encounters duplicate keys?

**Comment by ZhanGHanG9991** [comment]
### Performance Statistics
- NewLRUCache is implemented by LinkedHashMap.
- LRUCache is implemented by original doubly linked list and HashMap.
- OldLRUCacheis implemented by original doubly linked list.

| Method | Cache Size / Data Size | NewLRUCache Time | LRUCache Time | OldLRUCache Time |
| :----: | :--------------------: | :--------------: | :-----------: | :--------------: |
| Random |          0.07          |      0.412s      |    0.819s     |      1.440s      |
| Random |          0.1           |      0.389s      |    0.765s     |      1.449s      |
| Random |          0.2           |      0.356s      |    0.721s     |      1.361s      |
| Random |          0.3           |      0.365s      |    0.601s     |      1.303s      |
| Random |          0.4           |      0.361s      |    0.622s     |      1.268s      |
| Random |          0.5           |      0.339s      |    0.490s     |      1.154s      |
| Random |          0.6           |      0.318s      |    0.422s     |      1.170s      |
| Random |          0.7           |      0.191s      |    0.320s     |      1.115s      |
| Random |          0.8           |      0.193s      |    0.323s     |      1.142s      |
| Random |          0.9           |      0.194s      |    0.319s     |      0.924s      |

| Method | Cache Size / Data Size | NewLRUCache Time | LRUCache Time | OldLRUCache Time |
| :----: | :--------------------: | :--------------: | :-----------: | :--------------: |
|  Scan  |          0.07          |      0.338s      |    0.979s     |      1.384s      |
|  Scan  |          0.1           |      0.330s      |    0.950s     |      1.383s      |
|  Scan  |          0.2           |      0.325s      |    0.822s     |      1.708s      |
|  Scan  |          0.3           |      0.307s      |    0.685s     |      1.720s      |
|  Scan  |          0.4           |      0.308s      |    0.547s     |      1.713s      |
|  Scan  |          0.5           |      0.285s      |    0.445s     |      1.724s      |
|  Scan  |          0.6           |      0.292s      |    0.443s     |      1.713s      |
|  Scan  |          0.7           |      0.289s      |    0.441s     |      1.714s      |
|  Scan  |          0.8           |      0.288s      |    0.442s     |      1.721s      |
|  Scan  |          0.9           |      0.293s      |    0.442s     |      1.722s      |

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/NewCache.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.transformation.datastructure;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public abstract class NewCache extends LinkedHashMap<Integer, Integer> {
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/NewCache.java:25]: Rename to Cache.

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/Cache.java [hunk 3]
```diff
@@ -19,6 +19,9 @@
 
 package org.apache.iotdb.db.mpp.transformation.datastructure;
 
+import java.util.HashMap;
+import java.util.Map;
+
 /** <b>Note: It's not thread safe.</b> */
 public abstract class Cache {
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/Cache.java:26]: This class could be removed.

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/OldCache.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.transformation.datastructure;
+
+/** <b>Note: It's not thread safe.</b> */
+public abstract class OldCache {
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/transformation/datastructure/OldCache.java:23]: This class could be removed.

**Review by lancelly — APPROVED** [review]
LGTM～

**Pull request merged**